pkgs/games/factorio/default.nix

```diff
@@ -4,80 +4,108 @@
     };
-
-    x32 = {
-      url = "https://www.factorio.com/get-download/${version}/alpha/linux32";
-      sha256 = "13h013ixyhv4rpvh0jv5jry3mrwv65v57nqn16bjh3hr8ip70lkq";
+    i386 = {
+      alpha = authenticatedFetch { inherit      url; sha256 = "10135rd9103x79i89p6fh5ssmw612012yyx3yyhb3nzl554zqzbm"; };
     };
   };
-in
 
-stdenv.mkDerivation rec {
-  name = "factorio-${version}";
+  configBaseCfg = ''
+    use-system-read-write-data-directories=false
+    [path]
+    read-data=$out/share/factorio/data/
+  '';
 
-  src = fetch variants.${arch};
+  updateConfigSh = ''
+    #! $SHELL
+    if [[ -e ~/.factorio/config.cfg ]]; then
+      # Config file exists, but may have wrong path.
+      # Try to edit it. I'm sure this is perfectly safe and will never go wrong.
+      sed -i 's|^read-data=.*|read-data=$out/share/factorio/data/|' ~/.factorio/config.cfg
+    else
+      # Config file does not exist. Phew.
+      install -D $out/share/factorio/config-base.cfg ~/.factorio/config.cfg
+    fi
+  '';
 
-  libPath = stdenv.lib.makeLibraryPath [
-    alsaLib
-    libX11
-    libXcursor
-    libXinerama
-    libXrandr
-    libXi
-    mesa_noglu
-  ];
+in
+
+stdenv.mkDerivation rec {
+  name = "factorio-${releaseType}-${version}";
+
+  src = fetch.${arch.inTar}.${releaseType};
+
+  libPath = stdenv.lib.makeLibraryPath (
+    optionals (! isHeadless) [
+      alsaLib
+      libX11
+      libXcursor
+      libXinerama
+      libXrandr
+      libXi
+      mesa_noglu
+    ]
+  );
 
   buildInputs = [ makeWrapper ];
 
+  dontBuild = true;
+
+  # TODO detangle headless/normal mode wrapping, libs, etc.  test all urls 32/64/headless/gfx
   installPhase = ''
     mkdir -p $out/{bin,share/factorio}
-    cp -a bin/${arch}/factorio $out/bin/factorio.${arch}
-    cp -a doc-html data $out/share/factorio/
-
-    # Fortunately, Factorio already supports system-wide installs.
-    # Unfortunately it's a bit inconvenient to set the paths.
-    cat > $out/share/factorio/config-base.cfg <<EOF
-use-system-read-write-data-directories=false
-[path]
-read-data=$out/share/factorio/data/
-EOF
-
-    cat > $out/share/factorio/update-config.sh <<EOF
-if [[ -e ~/.factorio/config.cfg ]]; then
-  # Config file exists, but may have wrong path.
-  # Try to edit it. I'm sure this is perfectly safe and will never go wrong.
-  sed -i 's|^read-data=.*|read-data=$out/share/factorio/data/|' ~/.factorio/config.cfg
-else
-  # Config file does not exist. Phew.
-  install -D $out/share/factorio/config-base.cfg ~/.factorio/config.cfg
-fi
-EOF
-    chmod a+x $out/share/factorio/update-config.sh
-
+    cp -a data $out/share/factorio
+    cp -a bin/${arch.inTar}/factorio $out/bin/factorio
     patchelf \
       --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
-      $out/bin/factorio.${arch}
+      $out/bin/factorio
 
-    makeWrapper $out/bin/factorio.${arch} $out/bin/factorio \
+  '' + optionalString (! isHeadless) (''
+    mv $out/bin/factorio{,.${arch.inTar}}
```

Comment:
The braces? or moving the non-headless path? or just the optionalString part?

The other approach for separating the derivations that I considered was putting "// if (! isHeadless { ... } else {}" at the end.